emulator/src/kreator/preprocessor.rs

```diff
@@ -0,0 +1,532 @@
+            in_macro_declaration = true;
+            owned_line.clear();
+        }
+
+        // remove lines of macro declaration
+        if in_macro_declaration {
+            owned_line.clear();
+        }
+
+        // check lines for ENDM statement
+        if line.contains("ENDM") {
+            in_macro_declaration = false;
+            owned_line.clear();
+        }
+
+        for (macro_name, instructions) in &macros.0 {
+            if owned_line.contains(macro_name) {
+                let input_string = owned_line.split_once(macro_name).unwrap().1.trim();
+                let mut inputs: Vec<&str> = Vec::new();
+                for input in input_string.split(",") {
+                    inputs.push(input.trim());
+                }
+                let mut input_map: HashMap<String, String> = HashMap::new();
+                for (index, parameter) in macros.1.get(macro_name).unwrap().iter().enumerate() {
+                    let value = if index >= inputs.len() {
+                        String::new()
+                    } else {
+                        inputs[index].to_string()
+                    };
+                    input_map.insert(parameter.to_string(), value);
+                }
+                owned_line.clear();
+                for instruction in instructions {
+                    let mut line = instruction.to_string();
+                    for (variable, value) in &input_map {
+                        line = line.replace(&format!(" {} ", variable), &format!(" {} ", &value));
+                        line = line.replace(&format!(" {}", variable), &format!(" {}", &value));
+                    }
+                    macroless_code.push(line.trim().to_string());
+                }
+            }
+        }
+
+        if !owned_line.is_empty() {
+            macroless_code.push(owned_line.trim().to_string());
+        }
+    }
+    macroless_code.remove(macroless_code.len() - 1);
+    Ok(macroless_code)
+}
+
+fn eval_str(str: String) -> u16 {
+    let tokens = tokenize(str.to_string());
+    to_expression_tree(tokens).evaluate() as u16
+}
+
+fn get_labels(code: &Vec<String>) -> Result<HashMap<String, u16>, &'static str> {
+    let label_regex = Regex::new(LABEL_DECL).unwrap();
+    let reserved_names = vec![
+        "STC", "CMC", "INR", "DCR", "CMA", "DAA", "NOP", "MOV", "STAX", "LDAX", "ADD", "ADC",
+        "SUB", "SBB", "ANA", "XRA", "ORA", "CMP", "RLC", "RRC", "RAL", "RAR", "PUSH", "POP", "DAD",
+        "INX", "DCX", "XCHG", "XTHL", "SPHL", "LXI", "MVI", "ADI", "ACI", "SUI", "SBI", "ANI",
+        "XRI", "ORI", "CPI", "STA", "LDA", "SHLD", "LHLD", "PCHL", "JMP", "JC", "JNC", "JZ", "JNZ",
+        "JP", "JM", "JPE", "JPO", "CALL", "CC", "CNC", "CZ", "CNZ", "CP", "CM", "CPE", "CPO",
+        "RET", "RC", "RNC", "RZ", "RNZ", "RM", "RP", "RPE", "RPO", "RST", "EI", "DI", "IN", "OUT",
+        "HLT", "ORG", "EQU", "SET", "END", "IF", "ENDIF", "MACRO", "ENDM", "B", "C", "D", "H", "L",
+        "A", "SP", "PSW",
```

Comment:
I think trailing comments are unaesthetic, but its your call wether to remove it or not